src/solvers/strings/string_constraint_instantiation.cpp

```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
 /// \file
 /// Defines related function for string constraints.
 
+#include <unordered_set>
+#include <algorithm>
+#include <util/expr_iterator.h>
 #include "string_constraint_instantiation.h"
 
+/// look for the symbol and return true if it is found
+/// \param index: an index expression
+/// \param qvar: a symbol expression
+/// \return a Boolean
+static bool contains(const exprt &index, const symbol_exprt &qvar)
+{
+  return std::find(index.depth_begin(), index.depth_end(), qvar) !=
+         index.depth_end();
+}
+
+/// Find indexes of `str` used in `expr` that contains `qvar`, for instance
+/// with arguments ``(str[k+1]=='a')``, `str`, and `k`, the function should
+/// return `k+1`.
+/// \param [in] expr: the expression to search
+/// \param [in] str: the string which must be indexed
+/// \param [in] qvar: the universal variable that must be in the index
+/// \return index expressions in `expr` on `str` containing `qvar`.
+static std::unordered_set<exprt, irep_hash>
+find_indexes(const exprt &expr, const exprt &str, const symbol_exprt &qvar)
+{
+  decltype(find_indexes(expr, str, qvar)) result;
+  auto index_str_containing_qvar = [&](const exprt &e) {
+    if(auto index_expr = expr_try_dynamic_cast<index_exprt>(e))
+    {
+      const auto &arr = index_expr->array();
+      const auto str_it = std::find(arr.depth_begin(), arr.depth_end(), str);
+      return str_it != arr.depth_end() && contains(index_expr->index(), qvar);
+    }
+    return false;
+  };
+
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &e) {
+    if(index_str_containing_qvar(e))
+      result.insert(to_index_expr(e).index());
+  });
+  return result;
+}
+
+std::map<exprt, int> map_representation_of_sum(const exprt &f)
+{
+  // number of time the leaf should be added (can be negative)
+  std::map<exprt, int> elems;
+
+  std::list<std::pair<exprt, bool>> to_process;
+  to_process.emplace_back(f, true);
+
+  while(!to_process.empty())
+  {
+    exprt cur = to_process.back().first;
+    bool positive = to_process.back().second;
+    to_process.pop_back();
+    if(cur.id() == ID_plus)
+    {
+      for(const auto &op : cur.operands())
+        to_process.emplace_back(op, positive);
+    }
+    else if(cur.id() == ID_minus)
+    {
+      to_process.emplace_back(cur.op1(), !positive);
+      to_process.emplace_back(cur.op0(), positive);
+    }
+    else if(cur.id() == ID_unary_minus)
+    {
+      to_process.emplace_back(cur.op0(), !positive);
+    }
+    else
+    {
+      if(positive)
+        elems[cur] = elems[cur] + 1;
+      else
+        elems[cur] = elems[cur] - 1;
+    }
+  }
+  return elems;
+}
+
+exprt
+sum_over_map(std::map<exprt, int> &m, const typet &type, bool negated)
+{
+  exprt sum = nil_exprt();
+  mp_integer constants = 0;
```

Comment:
Don't declare or create them earlier than necessary. Notably they won't be needed at all if `m.empty()`.